Arm/Memory/Separate.lean

```diff
@@ -115,6 +116,505 @@ example : mem_subset (BitVec.ofNat 64 (2^64 - 1)) 0#64 0#64 (BitVec.ofNat 64 (2^
+  simp only [mem_legal', Nat.reducePow] at h ⊢
+  omega
+
+/--
+`mem_separate' a an b bn` asserts that two memory regions [a..an) and [b..bn) are separate.
+Note that we use *half open* intervals.
+See also: Why numbering should start at zero (https://www.cs.utexas.edu/~EWD/ewd08xx/EWD831.PDF)
+-/
+structure mem_separate' (a : BitVec 64) (an : Nat) (b : BitVec 64) (bn : Nat) : Prop where
+  ha : mem_legal' a an
+  hb : mem_legal' b bn
+  h : a.toNat + an ≤ b.toNat ∨ a.toNat ≥ b.toNat + bn
+
+-- @[simp]
+theorem BitVec.not_le_eq_lt {a b : BitVec w₁} : (¬ (a ≤ b)) ↔ b < a := by
+  rw [BitVec.le_def, BitVec.lt_def]
+  omega
+
+/-#
+This is a theorem we ought to prove, which establishes the equivalence
+between the old and new defintions of 'mem_separate'.
+However, the proof is finicky, and so we leave it commented for now.
+-/
+/-
+theorem mem_separate_of_mem_separate' (h : mem_separate' a an b bn)
+    (ha' : a' = a + (BitVec.ofNat w₁ (an - 1) ) (hb' : b' = b + (BitVec.ofNat w₁ (bn - 1)))
+    (hlegala : mem_legal a an) (hlegalb: mem_legal b bn) :
+    mem_separate a a' b b' := by
+  simp [mem_separate]
+  simp [mem_overlap]
+  obtain ⟨ha, hb, hsep⟩ := h
+  simp [mem_legal'] at ha hb
+  subst ha'
+  subst hb'
+  apply Classical.byContradiction
+  intro hcontra
+  · sorry
+  · sorry
+-/
+
+/-- `mem_subset' a an b bn` witnesses that `[a..a+an)` is a subset of `[b..b+bn)`-/
+structure mem_subset' (a : BitVec 64) (an : Nat) (b : BitVec 64) (bn : Nat) : Prop where
+  ha : mem_legal' a an
+  hb : mem_legal' b bn
+  hstart : b ≤ a
+  hend : a.toNat + an ≤ b.toNat + bn
+
+theorem mem_subset'_refl (h : mem_legal' a an) : mem_subset' a an a an where
+  ha := h
+  hb := h
+  hstart := by simp [BitVec.le_def]
+  hend := by simp [BitVec.le_def]
+
+/--
+If `[a'..a'+an')` begins at least where `[a..an)` begins,
+and ends before `[a..an)` ends, and if `[a..an)` is a subset of `[b..bn)`,
+then `[a'..a'+an')` is a subset of `[b..bn)`.
+-/
+theorem mem_subset'_of_le_of_le {a' : BitVec 64} (h : mem_subset' a an b bn)
+  (ha' : a.toNat ≤ a'.toNat) (han' : a'.toNat + an' ≤ a.toNat + an) :
+  mem_subset' a' an' b bn where
+  ha := by
+    obtain ⟨ha, hb, hstart, hend⟩ := h
+    simp only [mem_legal', Nat.reducePow] at ha hb ⊢
+    simp_all only [BitVec.le_def]
+    omega
+  hb := h.hb
+  hstart := by
+    obtain ⟨ha, hb, hstart, hend⟩ := h
+    simp only [mem_legal', Nat.reducePow] at ha hb
+    simp_all only [BitVec.le_def]
+    omega
+  hend := by
+    obtain ⟨ha, hb, hstart, hend⟩ := h
+    simp only [mem_legal', Nat.reducePow] at ha hb
+    simp_all only [BitVec.le_def]
+    omega
+
+/--
+If `[a..an)` is a subset of `[b..bn)`,
+then `[a..an')` is a subset of `[b..bn)` if `an' ≤ an`.
+-/
+theorem mem_subset'_of_length_le (h : mem_subset' a an b bn)
+  (han' : an' ≤ an) : mem_subset' a an' b bn := by
+  apply mem_subset'_of_le_of_le h
+  · simp [BitVec.le_def]
+  · omega
+
+/-- if `[a..an) ≤ [b..bn)` and `[b..bn) ⟂ [c..cn)`, then `[a..an) ⟂ [c..cn)`. -/
```

Comment:
Minor nit: I don't think you've used `⟂` before, might be worth explaining the symbol?

Reply:
Done, added prose.